### PR TITLE
Move InterchainGasPaymaster from AbacusConnectionManager to AbacusConnectionClient

### LINKED ISSUE #531: Consider moving the stored InterchainGasPaymaster out of ACM and into ACC
* It's becoming more clear that applications will want to use a common AbacusConnectionManager, but they may want to operate their own relayer and therefore use their own InterchainGasPaymaster. Currently, the pointer to the InterchainGasPaymaster used by an app is stored in the AbacusConnectionManager, so an app looking to use their own InterchainGasPaymaster would need to deploy and manage their own AbacusConnectionManager :(
* The main reason why the InterchainGasPaymaster address is in the AbacusConnectionManager at the moment is because of the relationship between an InterchainGasPaymaster and an Outbox-- the InterchainGasPaymaster takes payments based off the leaf index of a message in a particular Outbox, so when upgrading an Outbox the InterchainGasPaymaster should be atomically swapped out to avoid any inconsistency. We should consider bringing the stored InterchainGasPaymaster in the AbacusConnectionClient, and to pass the Outbox address to `InterchainGasPaymaster.payGasFor` so payments for messages can be differentiated based off their Outbox version.
* This issue encapsulates thinking through if this is the best solution, and making the smart contract changes
* Remove the setting of IGP in the Core Deployer

### PR DESCRIPTION
**Pull request by asaj**
This PR moves the `InterchainGasPaymaster` from `AbacusConnectionManager` to `AbacusConnectionClient`.

Decoupling the paymaster from the connection manager allows applications to share a connection manager while paying different relayers for their services.

`InterchainGasPaymaster` contracts are now able to serve as the payment address for multiple relayer binaries. These relayers may be configured to relay messages from different outboxes.

In the future, we should consider moving `InterchainGasPaymaster` and `AbacusConnectionManager` from `core` to `app`.

Other changes:
- Rename `TestXappConnectionClient` to `TestAbacusConnectionClient`
- Remove `interchainGasPaymaster` from SDK environments

Note @tkporter this may not play nicely with #504, as we no longer write an InterchainGasPaymaster to the rust configs. Open to suggestions for the best way to deal with this for now.

Fixes #531

### CODE REVIEW

File: solidity/core/interfaces/IInterchainGasPaymaster.sol [hunk 1]
```diff
@@ -2,17 +2,10 @@
...
- * @dev This is only intended for paying for messages sent via a specific
- * Outbox contract on the same source chain.
+ * @title InterchainGasPaymaster
+ * @notice Manages payments on a source chain to cover gas costs of relaying
+ * messages to destination chains.
  */
 interface IInterchainGasPaymaster {
-    /**
-     * @notice Deposits the msg.value as a payment for the proving & processing
-     * of a message on its destination chain.
```
Comment (yorhodes) [inline solidity/core/interfaces/IInterchainGasPaymaster.sol:14]: can we keep natspec here?

**Review by yorhodes — APPROVED** [review]

File: solidity/app/contracts/AbacusConnectionClient.sol [hunk 2]
```diff
@@ -13,7 +13,13 @@ abstract contract AbacusConnectionClient is OwnableUpgradeable {
     // ============ Mutable Storage ============
 
     IAbacusConnectionManager public abacusConnectionManager;
-    uint256[49] private __GAP; // gap for upgrade safety
+    // Interchain Gas Paymaster contract. The off-chain processor associated with
```
Comment (tkporter) [inline solidity/app/contracts/AbacusConnectionClient.sol:16]: should we s/processor/relayer?

Also wondering if we should make it clear in here that the interchain gas paymaster may present a reentrancy vector so the implementation should be confirmed by the application to not be able to re-enter. Also happy for this to just be a docs thing

File: solidity/core/contracts/AbacusConnectionManager.sol [hunk 3]
```diff
@@ -42,12 +36,6 @@ contract AbacusConnectionManager is IAbacusConnectionManager, Ownable {
      */
     event NewOutbox(address indexed outbox);
```
Comment (tkporter) [inline solidity/core/contracts/AbacusConnectionManager.sol:37]: would love for us to standardize event naming outside of this PR before we hit mainnet. We use NewFooBar and SetFooBar and should just choose one. I just filed https://github.com/abacus-network/abacus-monorepo/issues/560

File: solidity/core/interfaces/IInterchainGasPaymaster.sol [hunk 4]
```diff
@@ -2,17 +2,10 @@
 pragma solidity >=0.6.11;
 
 /**
- * @title IInterchainGasPaymaster
- * @notice An interface to pay source chain native tokens to cover the gas costs
- * of proving & processing messages on destination chains.
- * @dev This is only intended for paying for messages sent via a specific
- * Outbox contract on the same source chain.
+ * @title InterchainGasPaymaster
```
Comment (tkporter) [inline solidity/core/interfaces/IInterchainGasPaymaster.sol:5]: `IInterchainGasPaymaster`?

**Review by tkporter — APPROVED** [review]
Between the removal of the `interchainGasPaymaster` address in the config and the `GasPayment` event interface changing, like you said merging as is will once again put the contracts on main out of sync with the agents.
How about I make a new branch real quick with the agent changes required to handle this, make a PR to this branch, and then we can merge to main without breaking anything?

**Comment by tkporter** [comment]
PR with agent changes: https://github.com/abacus-network/abacus-monorepo/pull/562

File: solidity/core/contracts/AbacusConnectionManager.sol (hunk 3 above)
Comment (asaj) [reply solidity/core/contracts/AbacusConnectionManager.sol:37]: Yes please, agreed.

**Pull request merged**